### PR TITLE
[DPE-2167] Added admin extra user role

### PR DESCRIPTION
**Pull request by marceloneppel**
## Issue
There is no `admin` extra user role that can be requested by client application charms.

## Solution
Add only read permissions to the special `admin` role. The permission to create anything in the public schema of the `postgres` database was removed.

Also add to it the permission to create and drop databases (except the system database `postgres`).

While reviewing, the following files can be ignored (they are generated):
* `poetry.lock`
* `requirements.txt`
* `tests/integration/ha_tests/application-charm/requirements.txt`
* `tests/integration/new_relations/application-charm/requirements.txt`

We disabled the hashes temporarily on all the `requirements.txt` files due to [canonical/charmcraft#1179] (so we can pack the charms now). I will be reverted later.

Fixes https://github.com/canonical/postgresql-k8s-operator/issues/192.

### CODE REVIEW

File: lib/charms/postgresql_k8s/v0/postgresql.py [hunk 1]
```diff
@@ -178,28 +178,33 @@ def create_user(
...
+                    admin_role = "admin" in extra_user_roles
+                    roles = [role[0] for role in cursor.fetchall() if role[0] != "admin"]
+                    privileges = {
                         extra_user_role
                         for extra_user_role in extra_user_roles
-                        if extra_user_role not in roles
-                    ]
+                        if extra_user_role not in roles and extra_user_role != "admin"
+                    }
+                    if "SUPERUSER" in map(str.upper, privileges):
```
Comment (dragomirp) [inline lib/charms/postgresql_k8s/v0/postgresql.py:195]: I think PGB is using superuser. We should change that to admin after this change.

**Review by dragomirp — APPROVED** [review]

File: lib/charms/postgresql_k8s/v0/postgresql.py (hunk 1 above)
Comment (marceloneppel) [reply lib/charms/postgresql_k8s/v0/postgresql.py]: Good point. I reverted this part of the code as I realised that Landscape needs a superuser (for plpython extension, a requirement that may or may not be removed at some time, but also due to some schema changes that may happen in a landscape server upgrade).

Comment (taurus-forever) [reply lib/charms/postgresql_k8s/v0/postgresql.py]: 1) we should document this properly. JFMI, which extra_user_role used by pgbouncer?
2) re: `plpython`. We will have to hide it for now (c) Mohamed. Security concerns. This is why it was missing on mid-sync presentation. Mohamed will create a ticket to discuss and hide in documentation for now. No rush necessary today!

**Review by taurus-forever — APPROVED** [review]
Thank you!
What about to improve https://charmhub.io/postgresql-k8s/docs/e-users to expand "Relation/integration users" with description of extra_user_roles  (internal and allowed o be used "admin". BTW, if no `extra-user-roles` requested, how do we call this "role"? :-) )

File: lib/charms/postgresql_k8s/v0/postgresql.py [hunk 2]
```diff
@@ -129,7 +129,7 @@ def create_database(self, database: str, user: str) -> None:
                     sql.Identifier(database)
                 )
             )
-            for user_to_grant_access in [user] + self.system_users:
+            for user_to_grant_access in [user, "admin"] + self.system_users:
```
Comment (delgod) [inline lib/charms/postgresql_k8s/v0/postgresql.py:132]: Is this line needed if we add `pg_read_all_data, pg_write_all_data` to the role?

File: lib/charms/postgresql_k8s/v0/postgresql.py [hunk 3]
```diff
@@ -178,28 +178,28 @@ def create_user(
...
                 # Create or update the user.
                 cursor.execute(f"SELECT TRUE FROM pg_roles WHERE rolname='{user}';")
                 if cursor.fetchone() is not None:
                     user_definition = "ALTER ROLE {}"
                 else:
                     user_definition = "CREATE ROLE {}"
-                user_definition += (
-                    f"WITH LOGIN{' SUPERUSER' if admin else ''} ENCRYPTED PASSWORD '{password}'"
-                )
+                user_definition += f"WITH LOGIN{' SUPERUSER' if admin else ''} ENCRYPTED PASSWORD '{password}'{'IN ROLE admin CREATEDB CREATEROLE' if admin_role else ''}"
```
Comment (delgod) [inline lib/charms/postgresql_k8s/v0/postgresql.py:202]: 1. hm, is `CREATEROLE` needed for relation users? I think relation users with `extra_user_role=admin` shouldn't be able to create other users.
2. should we use `NOLOGIN` for `admin` role?
```suggestion
                user_definition += f"WITH {'NOLOGIN' if user == "admin" else 'LOGIN'}{' SUPERUSER' if admin else ''} ENCRYPTED PASSWORD '{password}'{'IN ROLE admin CREATEDB' if admin_role else ''}"
```

File: lib/charms/postgresql_k8s/v0/postgresql.py [hunk 4]
```diff
@@ -347,15 +347,21 @@ def set_up_database(self) -> None:
         """Set up postgres database with the right permissions."""
         connection = None
         try:
+            self.create_user(
+                "admin",
+                extra_user_roles="pg_read_all_data,pg_read_all_settings,pg_read_all_stats,pg_stat_scan_tables,pg_monitor,pg_signal_backend",
```
Comment (delgod) [inline lib/charms/postgresql_k8s/v0/postgresql.py:352]: I would limit privs by read and write all databases.
```suggestion
                extra_user_roles="pg_read_all_data,pg_write_all_data",
```

**Review by delgod — CHANGES_REQUESTED** [review]

File: lib/charms/postgresql_k8s/v0/postgresql.py (hunk 3 above)
Comment (marceloneppel) [reply lib/charms/postgresql_k8s/v0/postgresql.py]: Good catch! I fixed both situations on https://github.com/canonical/postgresql-k8s-operator/pull/201/commits/3b9591933d20ef01dddf3429181ac2a2f9607e6b.

File: lib/charms/postgresql_k8s/v0/postgresql.py (hunk 4 above)
Comment (marceloneppel) [reply lib/charms/postgresql_k8s/v0/postgresql.py]: Updated on https://github.com/canonical/postgresql-k8s-operator/pull/201/commits/3b9591933d20ef01dddf3429181ac2a2f9607e6b.

File: lib/charms/postgresql_k8s/v0/postgresql.py (hunk 2 above)
Comment (marceloneppel) [reply lib/charms/postgresql_k8s/v0/postgresql.py:132]: It's still needed after we simplify the privileges to only those two. At least the `CONNECT` privilege is needed (which is granted through this block).

File: lib/charms/postgresql_k8s/v0/postgresql.py (hunk 1 above)
Comment (marceloneppel) [reply lib/charms/postgresql_k8s/v0/postgresql.py]: PgBouncer currently uses `SUPERUSER` (to be able to create other superusers).

**Comment by marceloneppel** [comment]
> Thank you! What about to improve https://charmhub.io/postgresql-k8s/docs/e-users to expand "Relation/integration users" with description of extra_user_roles (internal and allowed o be used "admin". BTW, if no `extra-user-roles` requested, how do we call this "role"? :-) )

Good suggestion. I added the following text to both https://discourse.charmhub.io/t/charmed-postgresql-k8s-explanations-users/10843 and https://discourse.charmhub.io/t/charmed-postgresql-explanations-users/10798:

> When an application charm requests a new user through the relation/integration it can specify that the user should have the `admin` role in the `extra_user_roles` field. The `admin` role enables the new user to read and write to all databases (for the `postgres` system database it can only read data) and also to create and delete non-system databases.

**Review by delgod — APPROVED** [review]

**Pull request merged**